---
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, eval = T, results = 'hide',
                      fig.align = 'center', fig.width = 8, fig.height = 8)
knitr::opts_knit$set(root.dir = here::here())
```
```{r}
library(tidyverse)
library(cowplot)
library(DT)
library(knitr)
library(kableExtra)
source("R/themes.R")
```

## GWAS 
```{r}
gwas <- readRDS("results/mod_gwas.RDS")
```

### PCA-Based GWAS 
```{r}

bcorrectline <- -log10(0.05/nrow(gwas$gwas_modelret_pca))

gwas$gwas_modelret_pca %>% 
  dplyr::mutate(neglogp = -log10(pvalue),
                locinum = as.numeric(as.factor(loci))) %>% 
  ggplot() + 
  geom_point(aes(x=locinum, y = neglogp, color = neglogp)) +
  geom_hline(yintercept = bcorrectline, color = "red", linetype = "dashed") +
  labs(
    title = "GWAS with PCA-Covariate Matrix",
    x = "Genes",
    y = "-log(pvalue)") +
  scale_fill_viridis_c() +
  plot_theme +
  theme(
    plot.title = element_text(family = "Helvetica", face = "bold", hjust = 0.5, size = 12),
     axis.title.y = element_text(family = "Helvetica", hjust = 0.5, size = 11),
    axis.text.y = element_text(family = "Helvetica", hjust = 0.5, size = 8),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "none"
  )


```

#### Table Overview for Potentially Significant Loci 
```{r,  results = 'show'}
rosetta <- readr::read_csv("~/Documents/Github/P-REALM/data/derived/prealm_ret/bactopia-runs/pangenome-20251118-225922/panaroo/gene_presence_absence.csv")
rosetta <- rosetta[,1:3] %>% 
  magrittr::set_colnames(c("loci", "Non-unique Gene name", "Annotation"))
modelret_pca_rosetta <- dplyr::left_join(gwas$gwas_modelret_pca, rosetta, by = "loci")
modelret_pca_rosetta %>% 
  dplyr::select(c("loci", "Annotation", "pvalue")) %>% 
  DT::datatable(.,
                rownames = F,
                extensions='Buttons',
                options = list(
                  searching = T,
                  pageLength = 10,
                  dom = 'Bfrtip',
                  autoWidth = TRUE,
                  buttons = c('csv')))
```


#### Bayesian Approach for better Structure Control Looking at the Significant Loci 
Remember, confidence intervals should not cross zero (at bare minimum).
```{r,  results = 'show'}
bayes_rosetta <- dplyr::left_join(gwas$bayesmodelret_pca, rosetta, by = "loci")
bayes_rosetta %>% 
  dplyr::select(c("loci", "Annotation", "LCI", "UCI")) %>% 
  DT::datatable(.,
                rownames = F,
                extensions='Buttons',
                options = list(
                  searching = T,
                  pageLength = 10,
                  dom = 'Bfrtip',
                  autoWidth = TRUE,
                  buttons = c('csv')))

```



### Network GWAS 
```{r}

bcorrectline <- -log10(0.05/nrow(gwas$gwas_modelret_net))

gwas$gwas_modelret_net %>% 
  dplyr::mutate(neglogp = -log10(pvalue),
                locinum = as.numeric(as.factor(loci))) %>% 
  ggplot() + 
  geom_point(aes(x=locinum, y = neglogp, color = neglogp)) +
  geom_hline(yintercept = bcorrectline, color = "red", linetype = "dashed") +
  labs(
    title = "GWAS with Network-Covariate Matrix",
    x = "Genes",
    y = "-log(pvalue)") +
  scale_fill_viridis_c() +
  plot_theme +
  theme(
    plot.title = element_text(family = "Helvetica", face = "bold", hjust = 0.5, size = 12),
     axis.title.y = element_text(family = "Helvetica", hjust = 0.5, size = 11),
    axis.text.y = element_text(family = "Helvetica", hjust = 0.5, size = 8),
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "none"
  )


```
